In [1]:
from get_historical_prices import HistoricalPrices
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.style as style
style.use('seaborn-poster')
style.use('ggplot')
plt.rcParams['figure.figsize'] = (14, 8)

In [2]:
hp = HistoricalPrices(prices_dir='../prices')
hp.get_prices()
prices = hp.prices
prices.head()

100%|██████████| 966/966 [00:44<00:00, 21.92it/s]


high        low   adjclose      volume
ticker date                                                   
CSCO   2012-01-03  18.860001  18.480000  14.852622  41236600.0
       2012-01-04  19.000000  18.350000  15.139630  52927700.0
       2012-01-05  19.000000  18.670000  15.083825  37865300.0
       2012-01-06  19.000000  18.830000  15.028019  27796900.0
       2012-01-09  19.100000  18.790001  15.123683  37811500.0

In [3]:
pivoted_prices = prices.reset_index(level=0).pivot(columns='ticker', values='adjclose').dropna(axis=1)
print(pivoted_prices.shape)
pivoted_prices.head()

(1896, 796)


ticker              A         AA       AAL        AAP       AAPL        ABC  \
date                                                                          
2012-01-03  24.242914  20.893423  4.877474  68.042702  51.464207  33.846745   
2012-01-04  24.050198  21.391426  4.791737  69.637924  51.740776  33.962326   
2012-01-05  24.588482  21.187698  5.210895  70.081047  52.315197  34.220150   
2012-01-06  24.854307  20.734968  5.334737  70.405983  52.862103  34.309063   
2012-01-09  25.505566  21.346159  5.449052  70.386292  52.778248  34.291283   

ticker           ABMD        ABT        ACC       ACGL  ...    XPO       XRAY  \
date                                                    ...                     
2012-01-03  18.250000  22.961781  31.544683  12.493333  ...  11.84  33.780727   
2012-01-04  18.110001  22.872713  31.221191  12.266666  ...  11.72  33.771179   
2012-01-05  18.270000  22.820084  31.439375  12.500000  ...  11.73  33.799828   
2012-01-06  18.139999  22.613623  31.311480  12.486667  ...  11.83  33.599327   
2012-01-09  18.209999  22.609575  31.176056  12.396667  ...  11.70  33.437008   

ticker            XRX        XYL           Y        YUM       ZBRA        ZG  \
date                                                                           
2012-01-03  17.453468  22.564072  280.063232  34.874527  35.720001  7.193639   
2012-01-04  17.347052  23.309027  280.200897  35.112701  35.450001  7.174930   
2012-01-05  17.240623  22.976940  281.105408  35.380634  35.400002  7.171812   
2012-01-06  17.240623  22.591002  281.626495  35.636677  35.110001  7.184284   
2012-01-09  17.219339  22.680748  280.151733  35.499722  34.950001  7.181166   

ticker           ZION  ZNGA  
date                         
2012-01-03  15.602491  9.45  
2012-01-04  15.946727  9.19  
2012-01-05  16.393312  8.91  
2012-01-06  16.411919  8.81  
2012-01-09  16.635208  8.00  

[5 rows x 796 columns]

In [4]:
returns = pivoted_prices.pct_change()
returns.drop(returns.index[:1], inplace=True)
print(returns.shape)
returns.head()

(1895, 796)


ticker             A        AA       AAL       AAP      AAPL       ABC  \
date                                                                     
2012-01-04 -0.007949  0.023835 -0.017578  0.023444  0.005374  0.003415   
2012-01-05  0.022382 -0.009524  0.087475  0.006363  0.011102  0.007591   
2012-01-06  0.010811 -0.021368  0.023766  0.004637  0.010454  0.002598   
2012-01-09  0.026203  0.029476  0.021429 -0.000280 -0.001586 -0.000518   
2012-01-10  0.018760  0.001060  0.000000  0.000840  0.003580  0.011408   

ticker          ABMD       ABT       ACC      ACGL  ...       XPO      XRAY  \
date                                                ...                       
2012-01-04 -0.007671 -0.003879 -0.010255 -0.018143  ... -0.010135 -0.000283   
2012-01-05  0.008835 -0.002301  0.006988  0.019022  ...  0.000853  0.000848   
2012-01-06 -0.007116 -0.009047 -0.004068 -0.001067  ...  0.008525 -0.005932   
2012-01-09  0.003859 -0.000179 -0.004325 -0.007208  ... -0.010989 -0.004831   
2012-01-10 -0.002746  0.004655  0.013514  0.006453  ...  0.005983  0.014277   

ticker           XRX       XYL         Y       YUM      ZBRA        ZG  \
date                                                                     
2012-01-04 -0.006097  0.033015  0.000492  0.006829 -0.007559 -0.002601   
2012-01-05 -0.006135 -0.014247  0.003228  0.007631 -0.001410 -0.000435   
2012-01-06  0.000000 -0.016797  0.001854  0.007237 -0.008192  0.001739   
2012-01-09 -0.001235  0.003973 -0.005237 -0.003843 -0.004557 -0.000434   
2012-01-10  0.002472  0.012268  0.003685  0.006206  0.009156  0.079027   

ticker          ZION      ZNGA  
date                            
2012-01-04  0.022063 -0.027513  
2012-01-05  0.028005 -0.030468  
2012-01-06  0.001135 -0.011223  
2012-01-09  0.013605 -0.091941  
2012-01-10  0.016220  0.066250  

[5 rows x 796 columns]

## Looking at a two stock portfolio

In [11]:
tick1 = returns.columns[0]
tick2 = returns.columns[1]
asset_return_1 = returns[tick1].rename('asset_return_tick1')
asset_return_2 = returns[tick2].rename('asset_return_tick2')
print(type(asset_return_2))
asset_return_df = pd.concat([asset_return_1, asset_return_2], axis=1)
asset_return_df.head()

<class 'pandas.core.series.Series'>


asset_return_tick1  asset_return_tick2
date                                              
2012-01-04           -0.007949            0.023835
2012-01-05            0.022382           -0.009524
2012-01-06            0.010811           -0.021368
2012-01-09            0.026203            0.029476
2012-01-10            0.018760            0.001060

## Factor returns
Let's make up a "factor" by taking an average of all stocks in our list.  You can think of this as an equal weighted index of the 796 stocks, kind of like a measure of the "market".  We'll also make another factor by calculating the median of all the stocks.  These are mainly intended to help us generate some data to work with.

Also note that we're setting axis=1 so that we calculate a value for each time period (row) instead of one value for each column (assets).

In [12]:
factor_return_1 = returns.mean(axis=1)
factor_return_2 = returns.median(axis=1)
print(factor_return_2.head())
print(type(factor_return_2))
factor_return_df = pd.DataFrame(data={'factor_1': factor_return_1, 'factor_2': factor_return_2})
factor_return_df.head()
                                

date
2012-01-04   -0.001819
2012-01-05    0.005265
2012-01-06   -0.001441
2012-01-09    0.003229
2012-01-10    0.012424
dtype: float64
<class 'pandas.core.series.Series'>


factor_1  factor_2
date                          
2012-01-04 -0.001575 -0.001819
2012-01-05  0.006741  0.005265
2012-01-06 -0.000634 -0.001441
2012-01-09  0.004070  0.003229
2012-01-10  0.014502  0.012424

## Factor exposures

Factor exposures refer to how "exposed" a stock is to each factor.

In [13]:
from sklearn.linear_model import LinearRegression

In [14]:
# computing the factor exposures for asset one
lr1 = LinearRegression()
print('for regression with sklearn we need the columns to be the vectors which is perfect for our FR df')
X = factor_return_df.values
print(X.shape)
y = asset_return_1.values
print(y.shape)
lr1.fit(X,y)
lr1.coef_

for regression with sklearn we need the columns to be the vectors which is perfect for our FR df
(1895, 2)
(1895,)


array([5.51148265e-05, 1.26285550e+00])

In [15]:
# computing the factor exposures for asset two
lr2 = LinearRegression()
print('for regression with sklearn we need the columns to be the vectors which is perfect for our FR df')
X = factor_return_df.values
print(X.shape)
y = asset_return_2.values
print(y.shape)
lr2.fit(X,y)
lr2.coef_

for regression with sklearn we need the columns to be the vectors which is perfect for our FR df
(1895, 2)
(1895,)


array([-2.55030128e-04,  1.45225029e+00])

In [16]:
print(f'factor exposures for asset 1 are {lr1.coef_}')
print(f'factor exposures for asset 2 are {lr2.coef_}')      

factor exposures for asset 1 are [5.51148265e-05 1.26285550e+00]
factor exposures for asset 2 are [-2.55030128e-04  1.45225029e+00]


## Variance of stock 1

Calculate the variance of stock 1.  
$\textrm{Var}(f_{1}) = \beta_{1,1}^2 \textrm{Var}(f_{1}) + \beta_{1,2}^2 \textrm{Var}(f_{2}) + 2\beta_{1,1}\beta_{1,2}\textrm{Cov}(f_{1},f_{2}) + \textrm{Var}(s_{1})$

In [23]:
stock_1_factor_1_exposure = lr1.coef_[0]
stock_1_factor_2_exposure = lr1.coef_[1]
common_return_1 = stock_1_factor_1_exposure * factor_return_1 + stock_1_factor_2_exposure * factor_return_2
specific_return_1 = asset_return_1 - common_return_1
print(specific_return_1.head())
specific_return_1.shape

date
2012-01-04   -0.005652
2012-01-05    0.015732
2012-01-06    0.012631
2012-01-09    0.022125
2012-01-10    0.003070
dtype: float64


(1895,)

In [30]:
factor_cov_matrix = np.cov(factor_return_1, factor_return_2, ddof=1)
print('shape of cov matrix: ', factor_cov_matrix.shape)
factor_1_var = factor_cov_matrix[0,0]
factor_2_var = factor_cov_matrix[1,1]
cov_f1_f2 = factor_cov_matrix[0,1]

# calculate the variance of specific return
var_s_1 = np.var(specific_return_1, ddof=1)

# calculate the variance of stock one in terms of factors and specific risk
var_asset_1 = (stock_1_factor_1_exposure**2 * factor_1_var) + (stock_1_factor_2_exposure**2 * factor_2_var) + \
              (2 * stock_1_factor_1_exposure * stock_1_factor_2_exposure * cov_f1_f2) + var_s_1

var_asset_1_hist = np.var(asset_return_1, ddof=1)

print(f'variance of stock one with factors: {var_asset_1:.8f}')
print(f'variance of stock one with historical: {var_asset_1_hist:.8f}')

shape of cov matrix:  (2, 2)
variance of stock one with factors: 0.00022406
variance of stock one with historical: 0.00022406


## Variance of stock 2

Calculate the variance of stock 2.  
$\textrm{Var}(f_{2}) = \beta_{2,1}^2 \textrm{Var}(f_{1}) + \beta_{2,2}^2 \textrm{Var}(f_{2}) + 2\beta_{2,1}\beta_{2,2}\textrm{Cov}(f_{1},f_{2}) + \textrm{Var}(s_{2})$

In [31]:
stock_2_factor_1_exposure = lr2.coef_[0]
stock_2_factor_2_exposure = lr2.coef_[1]
common_return_2 = stock_2_factor_1_exposure * factor_return_1 + stock_2_factor_2_exposure * factor_return_2
specific_return_2 = asset_return_2 - common_return_2
print(specific_return_2.head())
specific_return_2.shape

date
2012-01-04    0.026477
2012-01-05   -0.017169
2012-01-06   -0.019274
2012-01-09    0.024789
2012-01-10   -0.016978
dtype: float64


(1895,)

In [33]:
# We already calculated the cov of the two factors and the variances of the factors

# calculate the variance of specific return
var_s_2 = np.var(specific_return_2, ddof=1)

# calculate the variance of stock one in terms of factors and specific risk
var_asset_2 = (stock_2_factor_1_exposure**2 * factor_2_var) + (stock_2_factor_2_exposure**2 * factor_2_var) + \
              (2 * stock_2_factor_1_exposure * stock_2_factor_2_exposure * cov_f1_f2) + var_s_2

var_asset_2_hist = np.var(asset_return_2, ddof=1)

print(f'variance of stock one with factors: {var_asset_2:.8f}')
print(f'variance of stock one with historical: {var_asset_2_hist:.8f}')

variance of stock one with factors: 0.00052981
variance of stock one with historical: 0.00052989


## Using matrices

Create matrices $\mathbf{F}$, $\mathbf{B}$ and $\mathbf{S}$, where  
$\mathbf{F}= \begin{pmatrix}
\textrm{Var}(f_1) & \textrm{Cov}(f_1,f_2) \\ 
\textrm{Cov}(f_2,f_1) & \textrm{Var}(f_2) 
\end{pmatrix}$
is the covariance matrix of factors,  

$\mathbf{B} = \begin{pmatrix}
\beta_{1,1}, \beta_{1,2}\\ 
\beta_{2,1}, \beta_{2,2}
\end{pmatrix}$ 
is the matrix of factor exposures, and  

$\mathbf{S} = \begin{pmatrix}
\textrm{Var}(s_i) & 0\\ 
0 & \textrm{Var}(s_j)
\end{pmatrix}$
is the matrix of specific variances.  

Then calculate $\mathbf{BFB}^T$

In [36]:
F = factor_cov_matrix # already calculated above
F

array([[ 1.23216945e+00, -6.47137449e-06],
       [-6.47137449e-06,  6.53805338e-05]])

In [39]:
B = np.array([lr1.coef_, lr2.coef_])
B

array([[ 5.51148265e-05,  1.26285550e+00],
       [-2.55030128e-04,  1.45225029e+00]])

In [45]:
S = np.diag([var_s_1, var_s_2])
S

array([[0.00011979, 0.        ],
       [0.        , 0.00039192]])

In [46]:
cov_assets = B.dot(F).dot(B.T) + S
cov_assets

array([[0.00022406, 0.00011989],
       [0.00011989, 0.00052989]])

In [47]:
# comparing to hist data:
np.cov(asset_return_1, asset_return_2, ddof=1)

array([[0.00022406, 0.00012733],
       [0.00012733, 0.00052989]])